### ISSUE TITLE
`borrow_as_ptr` triggers even when end result is reborrowed as `&mut`

### ISSUE THREAD

**Issue by Dexus0** [rust-lang/rust-clippy#11693]
### Summary

If the lint triggers when the pointer's reborrowed to a `&mut` it seems to me like it increases the chance of accidentally invoking undefined behavior, as `ptr::addr_of_mut!()`'s primary purpose seems to be for circumventing the restrictions of `&mut`, restrictions it's going to need anyway as it will be cast to that by the end of the statement.

#### TL;DR
The lint is suggesting the removal of guarantees, even-though it needs them later.

### Lint Name

borrow_as_ptr

### Reproducer

I tried `cargo clippy -- -W clippy::pedantic` on this code:

```rust
use std::io::Error;
use std::path::Path;

type ID = usize;

#[allow(dead_code)]
fn read_id(path: &Path) -> Result<ID, Error> {
    use std::{fs::File, io::Read, mem::size_of};

    let mut id: ID = 0;

    match File::open(path)?
        .read_exact(unsafe { &mut *(&mut id as *mut ID).cast::<[u8; size_of::<ID>()]>() })
    {
        Ok(()) => Ok(id),
        Err(e) => Err(e),
    }
}
```

I saw this happen:

```
warning: borrow as raw pointer
  --> src/main.rs:93:36
   |
93 |         .read_exact(unsafe { &mut *(&mut id as *mut ID).cast::<[u8; size_of::<ID>()]>() })
   |                                    ^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::addr_of_mut!(id)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrow_as_ptr
   = note: `-W clippy::borrow-as-ptr` implied by `-W clippy::pedantic`
```

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

**Comment by asquared31415** [comment]
No, in this case `&mut *(addr_of_mut(id).cast::<...>())` has the same semantics as what you wrote.  There are no guarantees created by `&mut id as *mut ID`, in fact that code is safe, so it **must** not have any conditions.  `addr_of_mut!(id)`'s power comes from not *imposing* extra requirements in certain situations that don't apply here.  In this case the lint should be considered more of a readabilty suggestion rather than the correctness cases that don't apply here.

(also, this code doesn't need to use `unsafe`, consider reading into a `let mut buf = [0_u8; size_of::<usize>()];` and using `usize::from_ne_bytes(buf)`)
